### PR TITLE
Bird readiness checks

### PR DESCRIPTION
**Pull request by ozdanborne**
## Description

Add readiness checks for Bird.

Implemented as a parent readiness check to the felix readiness check, since kubernetes allows just 1 readiness check per container.

The result of a failing readiness check when using `kubectl describe` looks neat:
```
Events:
  Type     Reason                 Age               From                          Message
  ----     ------                 ----              ----                          -------
  Normal   SuccessfulMountVolume  3m                kubelet, turk-kubeadm-node-1  MountVolume.SetUp succeeded for volume "var-run-calico"
  Normal   SuccessfulMountVolume  3m                kubelet, turk-kubeadm-node-1  MountVolume.SetUp succeeded for volume "cni-bin-dir"
  Normal   SuccessfulMountVolume  3m                kubelet, turk-kubeadm-node-1  MountVolume.SetUp succeeded for volume "var-lib-calico"
  Normal   SuccessfulMountVolume  3m                kubelet, turk-kubeadm-node-1  MountVolume.SetUp succeeded for volume "cni-net-dir"
  Normal   SuccessfulMountVolume  3m                kubelet, turk-kubeadm-node-1  MountVolume.SetUp succeeded for volume "lib-modules"
  Normal   SuccessfulMountVolume  3m                kubelet, turk-kubeadm-node-1  MountVolume.SetUp succeeded for volume "calico-node-token-vwdkc"
  Normal   SuccessfulMountVolume  3m                kubelet, turk-kubeadm-node-1  MountVolume.SetUp succeeded for volume "etcd-certs"
  Normal   Pulling                3m                kubelet, turk-kubeadm-node-1  pulling image "gcr.io/tigera-dev/experimental/node:turk1"
  Normal   Pulled                 3m                kubelet, turk-kubeadm-node-1  Successfully pulled image "gcr.io/tigera-dev/experimental/node:turk1"
  Normal   Created                3m                kubelet, turk-kubeadm-node-1  Created container
  Normal   Started                3m                kubelet, turk-kubeadm-node-1  Started container
  Normal   Pulled                 3m                kubelet, turk-kubeadm-node-1  Container image "quay.io/calico/cni:v2.1.0-rc1" already present on machine
  Normal   Created                3m                kubelet, turk-kubeadm-node-1  Created container
  Normal   Started                3m                kubelet, turk-kubeadm-node-1  Started container
  Warning  Unhealthy              5s (x3 over 25s)  kubelet, turk-kubeadm-node-1  Readiness probe failed: Bird is not ready: Unestablished peers: Mesh_10_128_0_24, Mesh_10_128_0_35
```

## Todos

- [x] Tests
- [ ] Documentation
- [x] Release note

## Release Note

```release-note
Calico/node includes a new readiness check for bird which reports the state of its BGP connections. Kubernetes uses this new endpoint when checking Calico’s readiness.
```

### CODE REVIEW

**Comment by jpiper** [comment]
In my situation, we have a large cluster with n racks and rather than just configure the BGP peering to the ToRs for each peer individually, we just define n global peers, and on each node only 1 session will actually be established. This is easier to maintain than hundreds of separate BGPPeer objects. I've raised an issue about that separately (#1828)

I wonder if anyone else does this, and could be surprised when all their calico-node pods are coming up as "NotReady".

File: calico_node/filesystem/sbin/check_readiness [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    exit 1
+fi
+
+# Check bird GR status
+if [ $(/bin/birdcl -s /var/run/calico/bird.ctl show status | grep "Graceful restart recovery in progress") ]; then
+    echo "Bird is not ready: Graceful restart in progress."
+    exit 1
+fi
+
+echo "Bird and Felix are ready."
```
Comment (caseydavenport) [inline calico_node/filesystem/sbin/check_readiness]: "calico/node is ready" might be nicer.

Or, Bird -> BIRD

File: calico_node/filesystem/sbin/check_readiness [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    /bin/birdcl -s /var/run/calico/bird.ctl show protocols | tail -n +3 | while read -r line; do
+        protocol=`echo "$line" | awk '{print $2}'`
+        if [ "$protocol" != "BGP" ]; then
+            continue
+        fi
+
+        status=`echo $line | awk '{print $6}'`
+        if [ "$status" != "Established" ]; then
+            peer=`echo $line | awk '{print $1}'`
+            printf "$peer, "
```
Comment (caseydavenport) [inline calico_node/filesystem/sbin/check_readiness]: I'd drop the `,` here - the output will always end in a comma which is a bit odd.

File: calico_node/filesystem/sbin/check_readiness [hunk 3]
```diff
@@ -0,0 +1,46 @@
+#!/bin/sh
+
+# Check felix readiness
+if [ ! -z "$FELIX_CHECKREADINESS" ]; then
+    if [ ! /usr/bin/wget localhost:9099/readiness -O - ]; then
+        echo "Felix is not ready."
```
Comment (caseydavenport) [inline calico_node/filesystem/sbin/check_readiness]: It would be nice to have a consistent format for the messages users get to see.

e.g.

```
calico/node is ready
calico/node is not ready: felix is not ready
calico/node is not ready: BGP not established with 10.0.1.1 10.0.0.1
calico/node is not ready: BGP graceful restart in progress
```

File: calico_node/filesystem/sbin/check_readiness [hunk 4]
```diff
@@ -0,0 +1,46 @@
+#!/bin/sh
+
+# Check felix readiness
+if [ ! -z "$FELIX_CHECKREADINESS" ]; then
+    if [ ! /usr/bin/wget localhost:9099/readiness -O - ]; then
```
Comment (caseydavenport) [inline calico_node/filesystem/sbin/check_readiness]: When k8s calls the readiness endpoint, it checks that 200 <= RC <= 400. We should probably do the same here, but maybe wget is already doing that check and resolving it for us?

Also probably worth being explicit with a timeout via `-T` - by default wget will timeout after 900 seconds, which is a bit long for this check. Probably no more than 10 seconds given we configure k8s to call this script every 10s?

**Comment by ozdanborne** [comment]
>I wonder if anyone [...] could be surprised when all their calico-node pods are coming up as "NotReady".

Users will see the reason why calico is NotReady in the `kubectl describe pod` output (example above). Hopefully, they will recognize that they have dead / misconfigured bgp peers, and will make it to this doc: https://docs.projectcalico.org/v3.0/usage/decommissioning-a-node, or disable the readiness checks by removing them from the install manifest if the peers are intentionally disconnected (such as your case).

I'll keep an eye out for how much this bites the community and we can work on streamlining it more if its a problem.

File: calico_node/filesystem/sbin/check_readiness [hunk 5]
```diff
@@ -0,0 +1,46 @@
+#!/bin/sh
+
+# Check felix readiness
+if [ ! -z "$FELIX_CHECKREADINESS" ]; then
```
[the remark this answers is not in the document]
Comment (ozdanborne) [reply calico_node/filesystem/sbin/check_readiness]: >We probably shouldn't define new FELIX_ variables here
Absoultely agreed. I think I'm going crazy here - I could have sworn this was the variable I found in our manifests but looks like you're right - its HEALTHENABLED.

As for your other two points, if they want to disable readiness checks, they can just remove them from the manifest. Since there's no "service" running for these, its simpler than the Felix ones in that regard.

**Comment by caseydavenport** [comment]
> Users will see the reason why calico is NotReady in the kubectl describe pod output (example above). Hopefully, they will recognize that they have dead / misconfigured bgp peers, and will make it to this doc: https://docs.projectcalico.org/v3.0/usage/decommissioning-a-node, or disable the readiness checks by removing them from the install manifest if the peers are intentionally disconnected (such as your case).

We'll need to make this change super clear in the release notes and add really good documentation for when and how to use this feature effectively.

Looking further ahead we should:
- Fix the operational burden requiring people to configure non-existent BGP peers.
- Add a config option to exclude a given peer from this check for non-critical peers.

File: calico_node/tests/st/bgp/test_bgp.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+            # Allow network to converge
+            self.assert_true(workload_host1.check_can_ping(workload_host2.ip, retries=10))
+
+            # Check connectivity in both directions
+            self.assert_ip_connectivity(workload_list=[workload_host1,
+                                                       workload_host2],
+                                        ip_pass_list=[workload_host1.ip,
+                                                      workload_host2.ip])
+
+            # Check the BGP status on the BIRD/GoBGP host.
```
Comment (caseydavenport) [inline calico_node/tests/st/bgp/test_bgp.py]: This test doesn't work for GoBGP, right?

File: calico_node/tests/st/bgp/test_bgp.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+            def check_connected():
```
Comment (caseydavenport) [inline calico_node/tests/st/bgp/test_bgp.py]: This feels a bit over-the-top when it could just be two calls:

```python
def check_connected():
  check_bird_status(host1, [("node-to-node mesh", host2.ip, "Established")])
  check_bird_status(host2, [("node-to-node mesh", host1.ip, "Established")])
```

File: calico_node/tests/st/bgp/test_bgp.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+
+from tests.st.test_base import TestBase
+from tests.st.utils.docker_host import DockerHost, CLUSTER_STORE_DOCKER_OPTIONS
+from tests.st.utils.utils import check_bird_status, retry_until_success
+
+_log = logging.getLogger(__name__)
+_log.setLevel(logging.DEBUG)
+
+
+class TestReadiness(TestBase):
```
Comment (caseydavenport) [inline calico_node/tests/st/bgp/test_bgp.py]: I think we should at least have one more test where BGP is disabled (CALICO_NETWORKING_BACKEND=none), Felix readiness enabled.

It would be nice to have tests which check the other readiness conditions as well - e.g. graceful restart in progress

File: master/getting-started/kubernetes/troubleshooting.md [hunk 9]
```diff
@@ -53,4 +53,11 @@ However, if you do need to assign a particular address to a Pod, {{site.prodname
 - You can request an IP that is available in {{site.prodname}} IPAM using the `cni.projectcalico.org/ipAddrs` annotation.
 - You can request an IP using the `cni.projectcalico.org/ipAddrsNoIpam` annotation. Note that this annotation bypasses the configured IPAM plugin, and thus in most cases it is recommended to use the above annotation. 
 
-See the [Requesting a Specific IP address]({{site.baseurl}}/{{page.version}}/reference/cni-plugin/configuration#requesting-a-specific-ip-address) section in the CNI plugin reference documentation for more details.
\ No newline at end of file
+See the [Requesting a Specific IP address]({{site.baseurl}}/{{page.version}}/reference/cni-plugin/configuration#requesting-a-specific-ip-address) section in the CNI plugin reference documentation for more details.
+
+## calico/node Readiness probe failed: Bird is not ready: Unestablished peers
+
+- calico/node container may report "unready" with this message in Calico v3.1+.
```
[the remark this answers is not in the document]
Comment (caseydavenport) [reply master/getting-started/kubernetes/troubleshooting.md]: Ended up on an outdated commit, but don't miss this guy @ozdanborne

File: master/getting-started/kubernetes/troubleshooting.md [hunk 10]
```diff
@@ -53,4 +53,11 @@ However, if you do need to assign a particular address to a Pod, {{site.prodname
...
+- In most cases, this message means a particular peer is unreachable in the cluster
+- This message can also bring to light when a stale node exists in the datastore.
```
[the remark this answers is not in the document]
Comment (caseydavenport) [reply master/getting-started/kubernetes/troubleshooting.md]: Same - ended up on an outdated commit

File: calico_node/filesystem/sbin/check_readiness [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+        fi
+
+        status=`echo $line | awk '{print $6}'`
+        if [ "$status" != "Established" ]; then
+            peer=`echo $line | awk '{print $1}'`
+            printf "$peer "
+        fi
+    done
+}
+# Check that bird is enabled.
```
[the remark this answers is not in the document]
Comment (ozdanborne) [reply calico_node/filesystem/sbin/check_readiness]: That's what this is doing. `/etc/service/enabled/bird` is a more explicit indication of whether bird is running or not, since `CALICO_NETWORKING_BACKEND`  has [some interesting wildcard factors](https://github.com/projectcalico/calico/blob/master/calico_node/filesystem/etc/rc.local#L50-L51) as to when it runs bird.

[the remark this answers is not in the document]
Comment (caseydavenport) [reply calico_node/filesystem/sbin/check_readiness]: Right, I see.

But...we still try to talk to bird below this to check its status, even when it isn't enabled, so I think there's still an issue

[the remark this answers is not in the document]
Comment (caseydavenport) [reply calico_node/filesystem/sbin/check_readiness]: I don't really understand the point about wildcard factors - if `CALICO_NETWORKING_BACKEND=bird` and the file `/etc/service/enabled/bird` doesn't exist, then something has gone wrong and and we should fail the readiness check... right?

File: master/reference/node/configuration.md [hunk 12]
```diff
@@ -144,3 +144,33 @@ e.g.
...
+
+```yaml
+readinessProbe:
+  periodSeconds: 10
+  httpGet:
+    path: /readiness
+    port: 9099
+```
+
+**Bird (and Felix)**
```
Comment (caseydavenport) [inline master/reference/node/configuration.md]: BIRD is an acronym and should by all caps,

File: master/reference/node/configuration.md [hunk 13]
```diff
@@ -144,3 +144,33 @@ e.g.
 IP_AUTODETECTION_METHOD=interface=eth.*
 IP6_AUTODETECTION_METHOD=interface=eth.*
 ```
+
+### Node readiness
```
Comment (caseydavenport) [inline master/reference/node/configuration.md]: I'd organize this in reverse.

e.g.

### calico/node readiness

#### BGP and Policy readiness
Do this to enable readiness....

#### Policy-only readiness
Do _this_ to enable readiness...

File: calico_node/filesystem/sbin/check_readiness [hunk 14]
```diff
@@ -1,8 +1,32 @@
...
+for arg in "$@"; do
+    case ${arg} in
+        felix )
+            check_felix=true
+        ;;
+        bird )
+            check_bird=true
+        ;;
+        * )
+            echo "calico readiness checks are not properly configured"
```
Comment (caseydavenport) [inline calico_node/filesystem/sbin/check_readiness]: Should say "Invalid arg"

File: calico_node/filesystem/sbin/check_readiness [hunk 15]
```diff
@@ -1,8 +1,32 @@
...
+            exit 1
+    esac
+done
+
+if [ "$check_bird" = false ] && [ "$check_felix" = false ]; then
+    echo "calico readiness checks are not properly configured"
```
Comment (caseydavenport) [inline calico_node/filesystem/sbin/check_readiness]: Should say "Must enable at least one of: bird, felix"

File: calico_node/filesystem/sbin/check_readiness [hunk 16]
```diff
@@ -1,8 +1,32 @@
...
+    exit 1
+fi
 
 # Check felix readiness
-if [ "$FELIX_HEALTHENABLED" = "true" ]; then
+if [ "check_felix" = "true" ]; then
```
Comment (caseydavenport) [inline calico_node/filesystem/sbin/check_readiness]: This will always evaluate to false

File: calico_node/filesystem/sbin/check_readiness [hunk 17]
```diff
@@ -0,0 +1,78 @@
+#!/bin/sh
+# check_readiness [felix] [bird]
+#
+#   felix   check felix health
+#   bird    check bird help
```
Comment (caseydavenport) [inline calico_node/filesystem/sbin/check_readiness]: typo: help

File: calico_node/filesystem/sbin/check_readiness [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+    esac
+done
+
+if [ "$check_bird" = false ] && [ "$check_felix" = false ]; then
+    echo "calico readiness checks are not properly configured"
+    exit 1
+fi
+
+# Check felix readiness
+if [ "check_felix" = "true" ]; then
```
Comment (caseydavenport) [inline calico_node/filesystem/sbin/check_readiness]: $check_felix

**Comment by ozdanborne** [comment]
Needed to add support for ipv6 and handle args better, so rewrote this in go and submitted it over at #1841
